mpc/client/src/main.rs

```diff
@@ -0,0 +1,628 @@
+            *latest_status.write().await = status.clone();
+            if let Err(_) = tx_ui_clone.send(ui::Event::NewStatus(status)) {
+                break;
+            }
+        }
+    });
+    tokio::spawn(async move {
+        let tick_rate = Duration::from_millis(1000 / 60);
+        let mut last_tick = Instant::now();
+        loop {
+            // poll for tick rate duration, if no events, sent tick event.
+            let timeout = tick_rate.saturating_sub(last_tick.elapsed());
+            if event::poll(timeout).unwrap() {
+                match event::read().unwrap() {
+                    event::Event::Key(key) => tx_ui.send(ui::Event::Input(key)).unwrap(),
+                    event::Event::Resize(_, _) => tx_ui.send(ui::Event::Resize).unwrap(),
+                    _ => {}
+                };
+            }
+            if last_tick.elapsed() >= tick_rate {
+                if let Err(_) = tx_ui.send(ui::Event::Tick) {
+                    break;
+                }
+                last_tick = Instant::now();
+            }
+        }
+    });
+}
+
+#[tokio::main]
+async fn main() -> Result<(), DynError> {
+    if let Err(_) = tokio::fs::metadata(ZKGM_DIR).await {
+        tokio::fs::create_dir(ZKGM_DIR).await?;
+    }
+    let status = Arc::new(RwLock::new(Status::Idle));
+    let lock = Arc::new(AtomicBool::new(false));
+    let (tx_status, rx_status) = broadcast::channel(64);
+    let graceful = GracefulShutdown::new();
+    let status_clone = status.clone();
+    let token = CancellationToken::new();
+    let token_clone = token.clone();
+    let tx_status_clone = tx_status.clone();
+    let handle = tokio::spawn(async move {
+        let addr = SocketAddr::from(([0, 0, 0, 0], 0x1337));
+        let listener = TcpListener::bind(addr).await.unwrap();
+        loop {
+            tokio::select! {
+                Ok((stream, _)) = listener.accept() => {
+                    let io = TokioIo::new(stream);
+                    let status_clone = status_clone.clone();
+                    let tx_status_clone = tx_status_clone.clone();
+                    let lock_clone = lock.clone();
+                    let conn = hyper::server::conn::http1::Builder::new().serve_connection(
+                        io,
+                        service_fn(move |req| {
+                            handle(
+                                lock_clone.clone(),
+                                tx_status_clone.clone(),
+                                status_clone.clone(),
+                                req,
+                            )
+                        }),
+                    );
+                    let fut = graceful.watch(conn);
+                    tokio::task::spawn(async move {
+                        let _ = fut.await;
+                    });
+                }
+                _ = token_clone.cancelled() => {
+                    break;
+                }
+            }
```

Comment:
this can be https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#method.run_until_cancelled